### PR TITLE
Bug fix/aardvark url encoding and decoding (percent)

### PR DESCRIPTION
**Pull request by Simran-B**
### Scope & Purpose

Attempt to fix URL encoding issues. Also make `npm run build` and `npm start` cross-platform.

- [x] :hankey: Bugfix (requires CHANGELOG entry)
- [ ] :book: CHANGELOG entry made

#### Backports:

- [ ] No backports required
- [ ] Backports required for: *(Please specify versions and link PRs)*

#### Related Information

#13187

- [ ] Maintainer README

### Testing & Verification

- [ ] This change is a trivial rework / code cleanup without any test coverage.
- [ ] The behavior in this PR was *manually tested*

### Documentation

- [ ] Added entry to *Release Notes*

### CODE REVIEW

**Comment by Simran-B** [comment]
There are several warnings when I use `npm start`. This one is concerning:

```
../frontend/js/lib/moment.min.js
Module not found: Can't resolve './locale' in 'C:\Daten\ArangoDB\arangodb\js\apps\system\_admin\aardvark\APP\frontend\js\lib'
```

The test server UI does not load because of errors, and at least some of them seem to be caused by the endpoints not returning the expected data. For example, `http://localhost:3000/_db/_system/_admin/aardvark/config.js` should return:

```js
var frontendConfig = {"basePath":"","db":"_system","isEnterprise":false,"authenticationEnabled":true,"ldapEnabled":false,"isCluster":false,"engine":"rocksdb","statisticsEnabled":true,"foxxStoreEnabled":true,"foxxApiEnabled":true,"minReplicationFactor":1,"maxReplicationFactor":10,"defaultReplicationFactor":1,"maxNumberOfShards":1000}
```

but instead, the UI HTML is served?!

<details><summary>Error log</summary><pre><code>ReferenceError: frontendConfig is not defined

databaseUrl
C:/Daten/ArangoDB/arangodb/js/apps/system/_admin/aardvark/APP/frontend/js/arango/arango.js:950
  947 | if (!databaseName) {
  948 |   databaseName = '_system';
  949 | 
> 950 |   if (frontendConfig && frontendConfig.db) {
      | ^  951 |     databaseName = frontendConfig.db;
  952 |   }
  953 | } // react dev testing
View compiled

(anonymous function)
C:/Daten/ArangoDB/arangodb/js/apps/system/_admin/aardvark/APP/frontend/js/models/arangoCollectionModel.js:7
   4 | 
   5 | window.arangoCollectionModel = Backbone.Model.extend({
   6 |   idAttribute: 'name',
>  7 |   urlRoot: arangoHelper.databaseUrl('/_api/collection'),
     | ^   8 |   defaults: {
   9 |     id: '',
  10 |     name: '',
View compiled

../frontend/js/models/arangoCollectionModel.js
C:/Daten/ArangoDB/arangodb/js/apps/system/_admin/aardvark/APP/frontend/js/models/arangoCollectionModel.js:319
  316 |       return result;
  317 |     }
  318 |   });
> 319 | })();
View compiled

__webpack_require__
C:/Daten/ArangoDB/arangodb/js/apps/system/_admin/aardvark/APP/react/webpack/bootstrap:848
  845 | };
  846 | 
  847 | // Execute the module function
> 848 | modules[moduleId].call(module.exports, module, module.exports, hotCreateRequire(moduleId));
      | ^  849 | 
  850 | // Flag the module as loaded
  851 | module.l = true;
View compiled

fn
C:/Daten/ArangoDB/arangodb/js/apps/system/_admin/aardvark/APP/react/webpack/bootstrap:150
  147 | 		);
  148 | 		hotCurrentParents = [];
  149 | 	}
> 150 | 	return __webpack_require__(request);
      | ^  151 | };
  152 | var ObjectFactory = function ObjectFactory(name) {
  153 | 	return {
View compiled

webpackContext
C:/Daten/ArangoDB/arangodb/js/apps/system/_admin/aardvark/APP/frontend/js/models sync /^/.*$:49
requireAll
C:/Daten/ArangoDB/arangodb/js/apps/system/_admin/aardvark/APP/react/src/App.js:54
  51 | window.JST = {};
  52 | 
  53 | function requireAll(context) {
> 54 |   context.keys().forEach(context);
  55 |   _.each(context.keys(), function(key) {
  56 |     // detect and store ejs templates
  57 |     if (key.substring(key.length - 4, key.length) === '.ejs') {
View compiled

Module../src/App.js
C:/Daten/ArangoDB/arangodb/js/apps/system/_admin/aardvark/APP/react/src/App.js:118
  115 |   '../../frontend/js/views/'
  116 | ));
  117 | 
> 118 | requireAll(require.context(
  119 |   '../../frontend/js/models/'
  120 | ));
  121 | 
View compiled

__webpack_require__
C:/Daten/ArangoDB/arangodb/js/apps/system/_admin/aardvark/APP/react/webpack/bootstrap:848
  845 | };
  846 | 
  847 | // Execute the module function
> 848 | modules[moduleId].call(module.exports, module, module.exports, hotCreateRequire(moduleId));
      | ^  849 | 
  850 | // Flag the module as loaded
  851 | module.l = true;
View compiled

fn
C:/Daten/ArangoDB/arangodb/js/apps/system/_admin/aardvark/APP/react/webpack/bootstrap:150
  147 | 		);
  148 | 		hotCurrentParents = [];
  149 | 	}
> 150 | 	return __webpack_require__(request);
      | ^  151 | };
  152 | var ObjectFactory = function ObjectFactory(name) {
  153 | 	return {
View compiled

Module../src/index.tsx
http://localhost:3000/static/js/main.chunk.js:75802:62
__webpack_require__
C:/Daten/ArangoDB/arangodb/js/apps/system/_admin/aardvark/APP/react/webpack/bootstrap:848
  845 | };
  846 | 
  847 | // Execute the module function
> 848 | modules[moduleId].call(module.exports, module, module.exports, hotCreateRequire(moduleId));
      | ^  849 | 
  850 | // Flag the module as loaded
  851 | module.l = true;
View compiled

fn
C:/Daten/ArangoDB/arangodb/js/apps/system/_admin/aardvark/APP/react/webpack/bootstrap:150
  147 | 		);
  148 | 		hotCurrentParents = [];
  149 | 	}
> 150 | 	return __webpack_require__(request);
      | ^  151 | };
  152 | var ObjectFactory = function ObjectFactory(name) {
  153 | 	return {
View compiled

1
http://localhost:3000/static/js/main.chunk.js:77377:18
__webpack_require__
C:/Daten/ArangoDB/arangodb/js/apps/system/_admin/aardvark/APP/react/webpack/bootstrap:848
  845 | };
  846 | 
  847 | // Execute the module function
> 848 | modules[moduleId].call(module.exports, module, module.exports, hotCreateRequire(moduleId));
      | ^  849 | 
  850 | // Flag the module as loaded
  851 | module.l = true;
View compiled

checkDeferredModules
C:/Daten/ArangoDB/arangodb/js/apps/system/_admin/aardvark/APP/react/webpack/bootstrap:45
  42 | 	}
  43 | 	if(fulfilled) {
  44 | 		deferredModules.splice(i--, 1);
> 45 | 		result = __webpack_require__(__webpack_require__.s = deferredModule[0]);
     | ^  46 | 	}
  47 | }
  48 | 
View compiled

Array.webpackJsonpCallback [as push]
C:/Daten/ArangoDB/arangodb/js/apps/system/_admin/aardvark/APP/react/webpack/bootstrap:32
  29 | 	deferredModules.push.apply(deferredModules, executeModules || []);
  30 | 
  31 | 	// run deferred modules when all chunks ready
> 32 | 	return checkDeferredModules();
     | ^  33 | };
  34 | function checkDeferredModules() {
  35 | 	var result;
View compiled

(anonymous function)
http://localhost:3000/static/js/main.chunk.js:1:57


ReferenceError: frontendConfig is not defined

child.initialize
C:/Daten/ArangoDB/arangodb/js/apps/system/_admin/aardvark/APP/frontend/js/routers/router.js:219
  216 | initFinished: false,
  217 | initialize: function initialize() {
  218 |   // check frontend config for global conf settings
> 219 |   if (frontendConfig.isCluster === true) {
      | ^  220 |     this.isCluster = true;
  221 |   }
  222 | 
View compiled
▶ 2 stack frames were collapsed.

HTMLDocument.<anonymous>
C:/Daten/ArangoDB/arangodb/js/apps/system/_admin/aardvark/APP/frontend/js/routers/startApp.js:25
  22 |   }
  23 | });
  24 | $(document).ready(function () {
> 25 |   window.App = new window.Router();
     | ^  26 |   Backbone.history.start();
  27 |   window.App.handleResize();
  28 | }); // create only this one global event listener
View compiled
▶ 2 stack frames were collapsed.</code></pre></details>


Even if I set the env var `REACT_APP_ARANGODB_CONFIG_PATH` to a working server (`http://localhost:8529/_db/_system/_admin/aardvark/config.js`), there is still a problem:

```
TypeError: Cannot create property 'version' on string '<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <link rel="shortcut icon" href="favicon.ico">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta name="theme-color" content="#000000">
    <!-- <link rel="manifest" href="//manifest.json"> -->
    <title>ArangoDB Web Interface</title>
    <meta name="description" content="ArangoDB Admin Web Interface">
    <meta name="author" content="Heiko Kernbach, Michael Hackstein">
    <script src="http://localhost:8529/_db/_system/_admin/aardvark/config.js"></script>
  </head>
...

success
C:/Daten/ArangoDB/arangodb/js/apps/system/_admin/aardvark/APP/frontend/js/views/footerView.js:175
  172 | frontendConfig.version = data;
  173 | 
  174 | if (!frontendConfig.version.hasOwnProperty('version')) {
> 175 |   frontendConfig.version.version = 'N/A';
      | ^  176 | }
  177 | 
  178 | self.showServerStatus(true);
```

**Comment by Simran-B** [comment]
With my fixes, the following actions now work:
- Create a new document `test%25` and get redirected to it
- Click on this document in the collection viewer to get to the document
- Create a new edge `edge%25` with _from `collection_name/test%` and _to `collection_name/test%25`  and get redirected to it
- Click on the _from and _to links to get to the respective vertex
- Document key is decoded for breadcrumb (`test%25`, not `test%2525`)

One thing that does not work, but also doesn't seem directly related:
- Edit _from or _to field (the clickable links), e.g. `collection_name/test%` to `collection_name/test%25`, confirm and save, then click on the changed name - it navigates you to the previous vertex. It works if you re-open the document viewer for the edge however.

More things to test:
- All non-alphanumeric characters which are allowed in keys
- Other percent-encoded values

File: js/apps/system/_admin/aardvark/APP/frontend/js/views/documentView.js [hunk 1]
```diff
@@ -540,7 +534,7 @@
       $('#subNavigationBar .breadcrumb').html(
         '<a href="#collection/' + name[1] + '/documents/1">Collection: ' + (name[1].length > 64 ? name[1].substr(0, 64) + "..." : name[1]) + '</a>' +
         '<i class="fa fa-chevron-right"></i>' +
-        this.type.charAt(0).toUpperCase() + this.type.slice(1) + ': ' + name[2]
+        this.type.charAt(0).toUpperCase() + this.type.slice(1) + ': ' + decodeURI(name[2])
```
Comment (goedderz) [inline js/apps/system/_admin/aardvark/APP/frontend/js/views/documentView.js:537]: Isn't `decodeURIComponent` the correct call to make here? Plus, must the result be encoded regarding HTML entities?

Comment (Simran-B) [reply js/apps/system/_admin/aardvark/APP/frontend/js/views/documentView.js:537]: `encodeURIComponent()` is said to encode 11 characters more than `encodeURI()`: `# $ & + , / : ; = ? @`.

Allowed characters in fragment identifier are apparently: `! $ & ' ( ) * + , ; = - . _ ~ : @ / ?` (as well as alpha, digit and `%` with two hex digits).

`#` seems like the only character that would require encoding from `encodeURIComponent()`, but we disallow it in document keys. So both encoding schemes should actually be fine, but they should obviously be used consistently.

In HTML attribute values, `'` needs to be encoded as `&#39;` or `&apos;` and `"` as `&#34;` or `&quot;`.
`&`, `<` and `>` need to be encoded as `&amp;`, `&lt;` and `&gt;` everywhere.

- https://stackoverflow.com/questions/75980/when-are-you-supposed-to-use-escape-instead-of-encodeuri-encodeuricomponent
- https://stackoverflow.com/questions/2849756/list-of-valid-characters-for-the-fragment-identifier-in-an-url
- https://tools.ietf.org/html/rfc3986#appendix-A
- https://stackoverflow.com/questions/7381974/which-characters-need-to-be-escaped-in-html

**Comment by Simran-B** [comment]
The issues seem to be fixed by #14700, including editing _from/_to and then navigating to the vertex.

**Pull request closed without merging**